pkg/ctlog/config.go

```diff
@@ -0,0 +1,324 @@
+	// Since we can have multiple Fulcio secrets, we need to construct a set
+	// of files containing them for the RootsPemFile. Names don't matter
+	// so we just call them fulcio-%
+	// What matters however is to ensure that the filenames match the keys
+	// in the configmap / secret that we construc so they get properly mounted.
+	rootPems := make([]string, 0, len(c.FulcioCerts))
+	for i := range c.FulcioCerts {
+		rootPems = append(rootPems, fmt.Sprintf("%sfulcio-%d", rootsPemFileDir, i))
+	}
+	keyDER, err := x509.MarshalPKIXPublicKey(c.PrivKey.Public())
+	if err != nil {
+		logging.FromContext(ctx).Panicf("Failed to marshal the public key: %v", err)
+	}
+	proto := configpb.LogConfig{
+		LogId:        c.LogID,
+		Prefix:       c.LogPrefix,
+		RootsPemFile: rootPems,
+		PrivateKey: mustMarshalAny(&keyspb.PEMKeyFile{
+			Path:     privateKeyFile,
+			Password: c.PrivKeyPassword}),
+		PublicKey:      &keyspb.PublicKey{Der: keyDER},
+		LogBackendName: "trillian",
+		ExtKeyUsages:   []string{"CodeSigning"},
+	}
+
+	multiConfig := configpb.LogMultiConfig{
+		LogConfigs: &configpb.LogConfigSet{
+			Config: []*configpb.LogConfig{&proto},
+		},
+		Backends: &configpb.LogBackendSet{
+			Backend: []*configpb.LogBackend{{
+				Name:        "trillian",
+				BackendSpec: c.TrillianServerAddr,
+			}},
+		},
+	}
+	marshalledConfig, err := prototext.Marshal(&multiConfig)
+	secrets, err := c.marshalSecrets()
+	if err != nil {
+		return nil, err
+	}
+	secrets[ConfigKey] = marshalledConfig
+	return secrets, nil
+}
+
+// MarshalSecrets returns a map suitable for creating a secret out of
+// containing the following keys:
+// private - CTLog private key, PEM encoded and encrypted with the password
+// public - CTLog public key, PEM encoded
+// fulcio-%d - For each fulcioCerts, contains one entry so we can support
+// multiple.
+func (c *CTLogConfig) marshalSecrets() (map[string][]byte, error) {
+	// Encode private key to PKCS#1 ASN.1 PEM.
+	block := &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(c.PrivKey),
+	}
+	// Encrypt the pem
+	block, err := x509.EncryptPEMBlock(rand.Reader, block.Type, block.Bytes, []byte(c.PrivKeyPassword), x509.PEMCipherAES256) // nolint
+	if err != nil {
+		return nil, fmt.Errorf("failed to encrypt private key: %w", err)
+	}
+
+	privPEM := pem.EncodeToMemory(block)
+	if privPEM == nil {
+		return nil, fmt.Errorf("failed to encode encrypted private key")
+	}
+	// Encode public key to PKCS#1 ASN.1 PEM.
+	pubPEM := pem.EncodeToMemory(
+		&pem.Block{
+			Type:  "RSA PUBLIC KEY",
+			Bytes: x509.MarshalPKCS1PublicKey(c.PrivKey.Public().(*rsa.PublicKey)),
+		},
+	)
+	data := map[string][]byte{
+		PrivateKey: privPEM,
+		PublicKey:  pubPEM,
+	}
+	for i, cert := range c.FulcioCerts {
+		fulcioKey := fmt.Sprintf("fulcio-%d", i)
+
+		// Fetch only root certificate from the chain
+		// TODO(Hayden): Is this what we should be adding, or the intermediates,
```

Comment:
This looks correct, it should only be the last certificate in the chain, the root.